core/tee/tee_time_generic.c

```diff
@@ -139,3 +139,21 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
 		return tee_time_ta_set_offs(uuid, &offs, false);
 	}
 }
+
+void tee_time_mdelay(uint32_t ms)
+{
+	TEE_Time prev, curr;
+	uint32_t ms_elapsed;
+
+	tee_time_get_sys_time(&prev);
+
+	do {
+		tee_time_get_sys_time(&curr);
+
+		if (curr.seconds == prev.seconds)
+			ms_elapsed = curr.millis - prev.millis;
+		else
```

Comment:
What? I don't get it.
And BTW the proposed implementation is wrong too, as it won't be able to handle `ms > 2000` in any case (not sure we'd like to do some busy wait for more that 2 seconds but unless we explicitly check for a maximum value it should work). @vchong I withdraw my R-b on this patch -- sorry for that.
Here is my proposal:
```C
void tee_time_mdelay(uint32_t ms)
{
        TEE_Time curr, delta, end;

        tee_time_get_sys_time(&curr);
        delta.seconds = ms / 1000;
        delta.millis = ms % 1000;
        TEE_TIME_ADD(curr, delta, end);

        while (TEE_TIME_LT(curr, end))
                tee_time_get_sys_time(&curr);
}
```